x/stakeibc/keeper/icacallbacks_undelegate.go

```diff
@@ -81,36 +79,70 @@ func (k Keeper) UndelegateCallback(ctx sdk.Context, packet channeltypes.Packet,
 	k.Logger(ctx).Info(utils.LogICACallbackStatusWithHostZone(chainId, ICACallbackID_Undelegate,
 		icacallbackstypes.AckResponseStatus_SUCCESS, packet))
 
-	// Update delegation balances
-	err := k.UpdateDelegationBalances(ctx, hostZone, undelegateCallback)
+	// Calculate the stTokens that should be burned from the batch and get the latest
+	// completion time from the ack response
+	nativeTokensUnbonded, stTokensToBurn := k.CalculateTokensFromBatch(undelegateCallback.SplitUndelegations)
+	unbondingTime, err := k.GetLatestUnbondingCompletionTime(ctx, ackResponse.MsgResponses)
 	if err != nil {
 		k.Logger(ctx).Error(fmt.Sprintf("UndelegateCallback | %s", err.Error()))
 		return err
 	}
 
-	// Get the latest transaction completion time (to determine the unbonding time)
-	latestCompletionTime, err := k.GetLatestCompletionTime(ctx, ackResponse.MsgResponses)
+	// Update delegation balances on the validators and host zone
+	err = k.UpdateDelegationBalances(ctx, hostZone, undelegateCallback)
 	if err != nil {
 		k.Logger(ctx).Error(fmt.Sprintf("UndelegateCallback | %s", err.Error()))
 		return err
 	}
 
-	// Burn the stTokens
-	stTokenBurnAmount, err := k.UpdateHostZoneUnbondings(ctx, *latestCompletionTime, chainId, undelegateCallback)
-	if err != nil {
+	// Update the accounting on the host zone unbondings
+	if err := k.UpdateHostZoneUnbondingsAfterUndelegation(
+		ctx,
+		chainId,
+		undelegateCallback.EpochUnbondingRecordIds,
```

Comment:
since the messages span multiple records, this will actually always be multiple. 

In a normal case, we'd have 4 epoch unbonding records here. You can see the excalidraw for a visual, but for example, we might have:

EUR1, EUR2, EUR3, EUR4 
val1, val2, val3, .... , val100

And a batch might just be val1 ... val5, but it will have EUR 1,2,3, and 4 in the callback data

Reply:
> this will actually always be multiple

Could you point me to where this is explained in the Excalidraw? My understanding is that whether they come from one record vs multiple would depend on the sorting or capacity / unbalanced ratio.

Reply:
![Screenshot 2024-05-23 at 11 39 08 AM](https://github.com/Stride-Labs/stride/assets/19301583/0fc66262-b8d9-4fd2-b567-b75dbbacbb73)

Setup: Two EUR's, 6 undelegation messages in batches of 2 per ICA

Notice here EUR1 and EUR2 are passed into the callback data _for every ICA_. And then in the callback it just cascades through each in order. So eventually EUR1 will be zero but we'll still loop over it, there will just be nothing to decrement

Reply:
Consider this resolved on my end